nbconvert/preprocessors/execute.py

```diff
@@ -441,44 +400,56 @@ def _wait_for_reply(self, msg_id, cell=None):
                 # not our reply
                 continue
 
+    def _get_timeout(self, cell):
+        if self.timeout_func is not None:
+            timeout = self.timeout_func(cell)
+        else:
+            timeout = self.timeout
+
+        if not timeout or timeout < 0:
+            timeout = None
+        return timeout
+
     def run_cell(self, cell, cell_index=0):
-        msg_id = self.kc.execute(cell.source)
+        from tornado import ioloop
+        from jupyter_kernel_mgmt.client import ErrorInKernel
+
+        timeout = self._get_timeout(cell)
+        interrupt_timeout = None
+        if self.interrupt_on_timeout:
+            interrupt_timeout = timeout
+            timeout += 5
+
+        out_msgs = []
+
         self.log.debug("Executing cell:\n%s", cell.source)
-        exec_reply = self._wait_for_reply(msg_id, cell)
+        try:
+            reply = self.kc.execute_interactive(
+                cell.source,
+                output_hook=out_msgs.append,
+                timeout=timeout,
+                interrupt_timeout=interrupt_timeout,
+                idle_timeout=self.iopub_timeout,
+                raise_on_no_idle=self.raise_on_iopub_timeout,
+            )
+        except ioloop.TimeoutError:
+            raise TimeoutError("Cell execution timed out")
+        except ErrorInKernel as e:
+            reply = e.reply_msg
```

Comment:
API question: should the client object raise an exception if the reply has `status: 'error'`? At the moment it does, but based on this and similar code in `jupyter_kernel_test`, I'm inclined to change it to return the reply, whatever the status of that reply.